src/pybind/mgr/deepsea/module.py

```diff
@@ -0,0 +1,290 @@
+    # Note: this is pretty braindead and doesn't implement the full eventsource
+    # spec, but it *does* implement enough for us to listen to events from salt
+    # and potentially do something with them.
+    # TODO: How are we going to deal with salt-api dying, or mgr failing over,
+    #       or other unforeseen glitches when we're waiting for particular jobs
+    #       to complete?  What's to stop things falling through the cracks?
+    def _read_sse(self):
+        event = {}
+        try:
+            for line in self._event_response.iter_lines():
+                if line:
+                    line = line.decode('utf-8')
+                    colon = line.find(':')
+                    if colon > 0:
+                        k = line[:colon]
+                        v = line[colon+2:]
+                        if k == "retry":
+                            # TODO: find out if we need to obey this reconnection time
+                            self.log.warn("Server requested retry {}, ignored".format(v))
+                        else:
+                            event[k] = v
+                else:
+                    # Empty line, terminates an event.  Note that event['tag']
+                    # is a salt-api extension to SSE to avoid having to decode
+                    # json data if you don't care about it.  To get to the
+                    # interesting stuff, you want event['data'], which is json.
+                    # TODO: this should probably really be debug level
+                    self.log.info("Got event '{}'".format(str(event)))
+
+                    # If we actually wanted to do something with the event,
+                    # say, we want to notice that some long salt run has
+                    # finished, we'd call some notify method here (TBD).
+
+                    # If you want to have some fun, try
+                    # `ceph daemon mgr.$(hostname) config set debug_mgr 4/5`
+                    # then `salt '*' test.ping` on the master
+                    event = {}
+            self._set_last_failure_msg("SSE read terminated")
+        except Exception as ex:
+            self._set_last_failure_msg("SSE read failed: {}".format(str(ex)))
+
+        self._reading_events = False
+        self.event.set()
+
+
+    # _do_request(), _login() and _do_request_with_login() are an extremely
+    # minimalist form of the following, with notably terse error handling:
+    # https://bitbucket.org/openattic/openattic/src/ce4543d4cbedadc21b484a098102a16efec234f9/backend/rest_client.py?at=master&fileviewer=file-view-default
+    # https://bitbucket.org/openattic/openattic/src/ce4543d4cbedadc21b484a098102a16efec234f9/backend/deepsea.py?at=master&fileviewer=file-view-default
+    # rationale:
+    # - I needed slightly different behaviour than in openATTIC (I want the
+    #   caller to read the response, to allow streaming the salt-api event bus)
+    # - I didn't want to pull in 400+ lines more code into this presently
```

Comment:
It's already pulled: [dashboard/rest_client.py](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/rest_client.py). Does it make sense to reuse it?

Reply:
Immediately I'm inclined to leave this as-is (just because it's small and self-contained), but if we could somehow reuse dashboard/rest_client.py, then that means less duplicated functionality. I'm just not sure how to do that :-) I mean, that code would have to be moved out of the dashboard to some common utility library or something, right? Also, it would need a tweak so that do_request() could return the Response object, so _read_sse() could stream from that.

Reply:
Bringing rest_client up into a shared utility seems pretty reasonable to me